include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,357 @@
+// instructions to simplify working with dialects in scenarios requiring
+// association of dialect operations with certain values.
+template <typename ValueT> class OpMap final {
+  using DialectOpKey = std::pair<StringRef, bool>;
+
+  struct DialectOpKV final {
+    DialectOpKey Key;
+    ValueT Value;
+
+    bool operator==(const DialectOpKV &other) const {
+      return Key.first == other.Key.first && Key.second == other.Key.second &&
+             Value == other.Value;
+    }
+
+    bool operator==(const OpDescription &desc) const {
+      const bool isOverload =
+          desc.getKind() == OpDescription::Kind::DialectWithOverloads;
+      return Key.first == desc.getMnemonic() && Key.second == isOverload;
+    }
+  };
+
+public:
+  OpMap() = default;
+
+  // -------------------------------------------------------------
+  // Convenience constructor to initialize the OpMap from a set of
+  // OpDescription/Value pairs.
+  // -------------------------------------------------------------
+  OpMap(std::initializer_list<std::pair<OpDescription, ValueT>> vals) {
+    for (const std::pair<OpDescription, ValueT> &val : vals)
+      insertOrGet(val.first, val.second);
+  }
+
+  // -------------------------------------------------------------
+  // Comparison operator overloads.
+  // -------------------------------------------------------------
+  bool operator==(const OpMap &rhs) const {
+    if (m_dialectOps.size() != rhs.m_dialectOps.size())
+      return false;
+
+    if (m_coreOpcodes == rhs.m_coreOpcodes &&
+        m_intrinsics == rhs.m_intrinsics) {
+      // Do a lookup for each vector entry, since both LHS and RHS potentially
+      // are in different order.
+      for (const auto &dialectOp : rhs.m_dialectOps) {
+        if (std::find(m_dialectOps.begin(), m_dialectOps.end(), dialectOp) ==
+            m_dialectOps.end())
+          return false;
+      }
+
+      return true;
+    }
+
+    return false;
+  }
+
+  bool operator!=(const OpMap &rhs) const { return !(*this == rhs); }
+
+  // -------------------------------------------------------------
+  // contains checks if a given op is contained in any of the
+  // internal map containers.
+  // -------------------------------------------------------------
+
+  bool containsCoreOp(unsigned op) const { return m_coreOpcodes.contains(op); }
+
+  bool containsIntrinsic(unsigned op) const {
+    return m_intrinsics.contains(op);
+  }
+
+  // Check if the map contains an OpDescription created for a given dialect
+  // operation type.
+  template <typename OpT> bool contains() const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return contains(desc);
+  }
+
+  // Check if the map contains an op described by an OpDescription.
+  bool contains(const OpDescription &desc) const {
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsCoreOp(op) || containsIntrinsic(op))
+          return true;
```

Comment:
This needs to only check either `containsCoreOp` or `containsIntrinsic`, depending on the incoming `desc`.